crates/rpc/src/v03/method/get_events.rs

```diff
@@ -0,0 +1,893 @@
+    }
+}
+
+/// Append's pending events to `dst` based on the filter requirements and returns
+/// true if this was the last pending data i.e. `is_last_page`.
+async fn append_pending_events(
+    pending_data: &Option<PendingData>,
+    dst: &mut Vec<types::EmittedEvent>,
+    skip: usize,
+    amount: usize,
+    address: Option<ContractAddress>,
+    keys: Vec<std::collections::HashSet<EventKey>>,
+) -> bool {
+    let pending_block = match pending_data.as_ref() {
+        Some(data) => match data.block().await {
+            Some(block) => block,
+            None => return true,
+        },
+        None => return true,
+    };
+
+    let original_len = dst.len();
+
+    let pending_events = pending_block
+        .transaction_receipts
+        .iter()
+        .flat_map(|receipt| {
+            receipt
+                .events
+                .iter()
+                .zip(std::iter::repeat(receipt.transaction_hash))
+        })
+        .filter(|(event, _)| match address {
+            Some(address) => event.from_address == address,
+            None => true,
+        })
+        .filter(|(event, _)| {
+            // check if filter is empty
+            if keys.iter().flatten().count() == 0 {
+                return true;
+            }
+
+            let keys_to_check = std::cmp::min(keys.len(), event.keys.len());
+
+            for (key, filter) in event
+                .keys
+                .iter()
+                .take(keys_to_check)
+                .zip(keys.iter().take(keys_to_check))
+            {
+                if !filter.contains(key) {
+                    return false;
+                }
+            }
+            true
+        })
+        .skip(skip)
+        // We need to take an extra event to determine is_last_page.
+        .take(amount + 1)
+        .map(|(event, tx_hash)| types::EmittedEvent {
+            data: event.data.clone(),
+            keys: event.keys.clone(),
+            from_address: event.from_address,
+            block_hash: None,
+            block_number: None,
+            transaction_hash: tx_hash,
+        });
+
+    dst.extend(pending_events);
+    let is_last_page = dst.len() <= (original_len + amount);
+    if !is_last_page {
+        dst.pop();
+    }
+
+    is_last_page
+}
+
+/// Continuation token is invalid if:
+/// 1. its value is nonzero (which means that it _actually_ points to some _next page_)
+/// 2. it yields an empty page
+///
+/// Unfortunately page retrieval has to be completed before the actual check can be done.
+fn check_continuation_token_validity(
+    continuation_token: Option<usize>,
+    events: &[types::EmittedEvent],
+) -> Result<(), GetEventsError> {
+    match continuation_token {
+        Some(token) if token > 0 && events.is_empty() => {
+            Err(GetEventsError::InvalidContinuationToken)
+        }
+        Some(_) | None => Ok(()),
+    }
+}
+
+fn next_continuation_token(
+    current_offset: usize,
+    chunk_size: usize,
+    is_last_page: bool,
```

Comment:
nit: I am no Clean Code fanatic, but in this particular case looks like `bool` argument confirms previous comment about `append_pending_events` and `is_last_page`.